### PR TITLE
abstract: make fibers cancellable

### PR DESCRIPTION
**Pull request by nshy**
We are going to finish all client (non-system) fibers in the process of Tarantool shutdown. First we cancel them and then wait for their finishing. So if the client fibers are not finished Tarantool shutdown is never finished too.

Currently some of queue fibers can not be finished and we got hang on integration testing of PR [1]. Let's fix it.

[1] PR with client fibers finishing on shutdown.
https://github.com/tarantool/tarantool/pull/9604

### CODE REVIEW

File: queue/abstract/driver/utubettl.lua [hunk 1]
```diff
@@ -169,7 +169,7 @@ local function utubettl_fiber(self)
         if box.info.ro == false then
             local stat, err = pcall(utubettl_fiber_iteration, self, processed)
 
-            if not stat and not err.code == box.error.READONLY then
+            if not stat and not (err.code == box.error.READONLY) then
```
Comment (LeonidVas) [inline queue/abstract/driver/utubettl.lua:172]: Please move this to a separate commit.

File: queue/abstract/queue_state.lua [hunk 2]
```diff
@@ -65,25 +65,21 @@ local function create_state_fiber(on_state_change_cb)
...
-                if rc then
-                    current = queue_state.states.STARTUP
-                    log.info('Queue state changed: STARTUP')
-                    -- Wait for maximum upstream lag * 2.
-                    fiber.sleep(max_lag() * 2)
-                    on_state_change_cb(current)
-                    current = queue_state.states.RUNNING
-                    log.info('Queue state changed: RUNNING')
-                end
+                box.ctl.wait_rw()
```
Comment (LeonidVas) [inline queue/abstract/queue_state.lua:68]: If I understand correctly, we can keep calling with `pcall` and in case of an error write a message to the log, check if `fiber` is canceled and then throw a new error with more additional context. If this is the case, I prefer to do the extra processing.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patch.
See several comments bellow.

File: queue/abstract/driver/utubettl.lua (hunk 1 above)
Comment (nshy) [reply queue/abstract/driver/utubettl.lua:172]: Done.

File: queue/abstract/queue_state.lua (hunk 2 above)
Comment (nshy) [reply queue/abstract/queue_state.lua]: Changed to:

```diff
--- a/queue/abstract/queue_state.lua
+++ b/queue/abstract/queue_state.lua
@@ -85,8 +85,12 @@ local function create_state_fiber(on_state_change_cb)
                     log.info('Queue state changed: WAITING')
                 end
             end
+            -- Handle server shutdown.
+            if not pcall(fiber.testcancel) then break end
         end
     end)
+
+    log.info('Finished queue state fiber')
 end

 -- Public methods.
```

**Review by LeonidVas — APPROVED** [review]
